rust/user-info-fetcher/src/main.rs

```diff
@@ -0,0 +1,262 @@
+        }
+        if let Some(tls_ca_cert_mount_path) = keycloak.tls.tls_ca_cert_mount_path() {
+            let mut buf = Vec::new();
+            File::open(tls_ca_cert_mount_path)
+                .await
+                .context(ReadCaCertSnafu)?
+                .read_to_end(&mut buf)
+                .await
+                .unwrap();
+            let ca_cert = reqwest::Certificate::from_pem(&buf).context(ParseCaCertSnafu)?;
+
+            client_builder = client_builder
+                .tls_built_in_root_certs(false)
+                .add_root_certificate(ca_cert);
+        }
+    }
+    let http = client_builder.build().context(ConstructHttpClientSnafu)?;
+
+    let user_info_cache = {
+        let crd::Cache { entry_time_to_live } = config.cache;
+        Cache::builder()
+            .name("user-info")
+            .time_to_live(*entry_time_to_live)
+            .build()
+    };
+    let app = Router::new()
+        .route("/user", post(get_user_info))
+        .with_state(AppState {
+            config,
+            http,
+            credentials,
+            user_info_cache,
+        });
+    axum::Server::bind(&"127.0.0.1:9476".parse().context(ParseListenAddrSnafu)?)
+        .serve(app.into_make_service())
+        .with_graceful_shutdown(shutdown_requested)
+        .await
+        .context(RunServerSnafu)
+}
+
+#[derive(Debug, Deserialize, PartialEq, Eq, Hash, Clone)]
+#[serde(rename_all = "camelCase", untagged)]
+enum UserInfoRequest {
+    UserInfoRequestById(UserInfoRequestById),
+    UserInfoRequestByName(UserInfoRequestByName),
+}
+
+#[derive(Debug, Deserialize, PartialEq, Eq, Hash, Clone)]
+#[serde(rename_all = "camelCase")]
+struct UserInfoRequestById {
+    id: String,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Eq, Hash, Clone)]
+#[serde(rename_all = "camelCase")]
+struct UserInfoRequestByName {
+    username: String,
+}
+
+#[derive(Serialize, Clone, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+struct UserInfo {
+    /// This might be null in case the id is not known (e.g. the backend does not have this info).
+    id: Option<String>,
+    /// This might be null in case the username is not known (e.g. the backend does not have this info).
+    username: Option<String>,
+    groups: Vec<String>,
+    custom_attributes: HashMap<String, Vec<String>>,
+}
+
+#[derive(Snafu, Debug)]
+#[snafu(module)]
+enum GetUserInfoError {
+    #[snafu(display("failed to get user information from Keycloak"))]
+    Keycloak { source: backend::keycloak::Error },
+}
+
+impl http_error::Error for GetUserInfoError {
+    fn status_code(&self) -> hyper::StatusCode {
+        // todo: the warn here loses context about the scope in which the error occurred, eg: stackable_opa_user_info_fetcher::backend::keycloak
+        // Also, we should make the log level (warn vs error) more dynamic in the backend's impl `http_error::Error for Error`
+        tracing::warn!(error = ?self, "Error while processing request");
```

Comment:
```suggestion
        tracing::warn!(error = &self as &dyn std::error::Error, "Error while processing request");
```

Reply:
I assume the `&` is not needed, as it results in:

```
211 |     fn status_code(&self) -> hyper::StatusCode {
    |                    - let's call the lifetime of this reference `'1`
...
215 |             error = &self as &dyn std::error::Error,
    |                     ^^^^^ cast requires that `'1` must outlive `'static`
```

I have removed the `?` and added the cast in ab945d2819816c1781d324ab04225a90fa19bdd2

Reply:
You're right, since `self` is already `&Self` here.